testnet/stacks-node/src/nakamoto_node/signer_coordinator.rs

```diff
@@ -0,0 +1,376 @@
+use crate::event_dispatcher::StackerDBChannel;
+use crate::nakamoto_node::stackerdb_listener::{StackerDBListener, EVENT_RECEIVER_POLL};
+use crate::neon::Counters;
+use crate::Config;
+
+/// The state of the signer database listener, used by the miner thread to
+/// interact with the signer listener.
+pub struct SignerCoordinator {
+    /// The private key used to sign messages from the miner
+    message_key: StacksPrivateKey,
+    /// Is this mainnet?
+    is_mainnet: bool,
+    /// The session for writing to the miners contract in the stackerdb
+    miners_session: StackerDBSession,
+    /// The total weight of all signers
+    total_weight: u32,
+    /// The weight threshold for block approval
+    weight_threshold: u32,
+    /// Interface to the StackerDB listener thread's data
+    stackerdb_comms: StackerDBListenerComms,
+    /// Keep running flag for the signer DB listener thread
+    keep_running: Arc<AtomicBool>,
+    /// Handle for the signer DB listener thread
+    listener_thread: Option<JoinHandle<()>>,
+}
+
+impl SignerCoordinator {
+    /// Create a new `SignerCoordinator` instance.
+    /// This will spawn a new thread to listen for messages from the signer DB.
+    pub fn new(
+        stackerdb_channel: Arc<Mutex<StackerDBChannel>>,
+        node_keep_running: Arc<AtomicBool>,
+        reward_set: &RewardSet,
+        burn_tip: &BlockSnapshot,
+        burnchain: &Burnchain,
+        message_key: StacksPrivateKey,
+        config: &Config,
+    ) -> Result<Self, ChainstateError> {
+        info!("SignerCoordinator: starting up");
+        let keep_running = Arc::new(AtomicBool::new(true));
+
+        // Create the stacker DB listener
+        let mut listener = StackerDBListener::new(
+            stackerdb_channel,
+            node_keep_running.clone(),
+            keep_running.clone(),
+            reward_set,
+            burn_tip,
+            burnchain,
+        )?;
+        let is_mainnet = config.is_mainnet();
+        let rpc_socket = config
+            .node
+            .get_rpc_loopback()
+            .ok_or_else(|| ChainstateError::MinerAborted)?;
+        let miners_contract_id = boot_code_id(MINERS_NAME, is_mainnet);
+        let miners_session = StackerDBSession::new(&rpc_socket.to_string(), miners_contract_id);
+
+        let mut sc = Self {
+            message_key,
+            is_mainnet,
+            miners_session,
+            total_weight: listener.total_weight,
+            weight_threshold: listener.weight_threshold,
+            stackerdb_comms: listener.get_comms(),
+            keep_running,
+            listener_thread: None,
+        };
+
+        // Spawn the signer DB listener thread
+        let listener_thread = std::thread::Builder::new()
+            .name("stackerdb_listener".to_string())
+            .spawn(move || {
+                if let Err(e) = listener.run() {
+                    error!("StackerDBListener: exited with error: {e:?}");
+                }
```

Comment:
Hmm.. yes, I think we should. I don't remember if there was some reason that I didn't do that originally... it seems like I did it on purpose, but I can't see why now 😬.

Reply:
I guess it is a little strange, since that error will only have any effect when the signer coordinator eventually joins with that listener thread. All it does then is log the error. So it seems like we would just log the error twice. So I guess it doesn't help anything to bubble up the error.